types/src/v2/request.rs

```diff
@@ -76,3 +76,75 @@ impl<'a> JsonRpcNotificationSer<'a> {
 		Self { jsonrpc: TwoPointZero, method, params }
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use serde_json::Value;
+
+	fn assert_request<'a>(request: JsonRpcRequest<'a>, id: Option<&str>, method: &str, params: Option<&str>) {
+		assert_eq!(request.jsonrpc, TwoPointZero);
+		assert_eq!(request.id.map(RawValue::get), id);
+		assert_eq!(request.method, method);
+		assert_eq!(request.params.map(RawValue::get), params);
+	}
+
+	/// Checks that we can serialize the object with or without non-mandatory fields.
+	#[test]
+	fn request_serialize_works() {
+		let method = "subtract";
+		let id = Id::Number(1); // It's enough to check one variant, since the type itself also has tests.
+		let params = JsonRpcParams::Array(vec![Value::Number(42.into()), Value::Number(23.into())]); // Same as above.
+		let test_vector = &[
+			// With all fields set.
+			(
+				r#"{"jsonrpc":"2.0","method":"subtract","id":1,"params":[42,23]}"#,
+				Some(id.clone()),
+				Some(params.clone()),
+			),
+			// Without ID field.
+			(r#"{"jsonrpc":"2.0","method":"subtract","id":null,"params":[42,23]}"#, None, Some(params)),
+			// Without params field
+			(r#"{"jsonrpc":"2.0","method":"subtract","id":1,"params":null}"#, Some(id), None),
+			// Without params and ID.
+			(r#"{"jsonrpc":"2.0","method":"subtract","id":null,"params":null}"#, None, None),
+		];
+
+		for (ser, id, params) in test_vector.iter().cloned() {
+			let request = serde_json::to_string(&JsonRpcCallSer {
+				jsonrpc: TwoPointZero,
+				method,
+				id: id.unwrap_or(Id::Null),
+				params: params.unwrap_or(JsonRpcParams::NoParams),
+			})
+			.unwrap();
+
+			assert_eq!(&request, ser);
+		}
+	}
+
+	/// Checks that we can deserialize the object with or without non-mandatory fields.
+	#[test]
+	fn request_deserialize_works() {
+		let method = "subtract";
+		let id = "1";
+		let params = "[42, 23]";
+		let test_vector = &[
+			// With all fields set.
+			(r#"{"jsonrpc":"2.0", "method":"subtract", "params":[42, 23], "id":1}"#, Some(id), Some(params)),
+			// Without ID field.
+			(r#"{"jsonrpc":"2.0", "method":"subtract", "params":[42, 23]}"#, None, Some(params)),
```

Comment:
not your fault but this shouldn't be parsed as a `JsonRpcRequest` instead should be parsed as `JsonRpcNotification`, #325 fixes this.

Reply:
Should I just remove it now, or wait until the linked PR is merged and implement a negative test?

Reply:
I merged it #325, do mind merging/rebasing to master? Otherwise, I can help with it too.

Sorry for introducing this merge conflicts.

Reply:
No problem. I've rebased to master, removed tests that became duplicated (tried to leave ones that are more comprehensive), and aligned my tests with the surrounding code.